[export] preserve metadata during nonstrict tracing

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118487
* #118486
* #118485
* #118484
* #118505
* __->__ #118483

Previously, nonstrict tracing would wipe metadata of graphmodules, because the wrapper class we're using was not detected as a graphmodule and thus meta preservation was not turned on

Differential Revision: [D53139354](https://our.internmc.facebook.com/intern/diff/D53139354/)

suo committed:
[export] preserve metadata during nonstrict tracing

Previously, nonstrict tracing would wipe metadata of graphmodules, because the wrapper class we're using was not detected as a graphmodule and thus meta preservation was not turned on

Differential Revision: [D53139354](https://our.internmc.facebook.com/intern/diff/D53139354/)

[ghstack-poisoned]

suo commented on torch/export/_trace.py:
```diff
@@ -551,9 +551,14 @@ def __init__(self, mod):
-                        flat_outs, out_spec = pytree.tree_flatten(
-                            self._export_root(*args, **kwargs)
-                        )
+                        if isinstance(self._export_root, torch.fx.GraphModule):
+                            with torch.fx.traceback.preserve_node_meta():
+                                tree_out = torch.fx.Interpreter(self._export_root).run(
+                                    *args, **kwargs
+                                )
+                        else:
+                            tree_out = self._export_root(*args, **kwargs)
```
I dunno, it's a pretty small and specific pattern so the helper function would not deliver a ton a value. It does show up both here and in aot_autograd though. 

I think the real issue isn't the code duplication—the more fundamental issue is that we have two different systems (aot_autograd and export) that think they're responsible for preserving metadata. And that our system for managing and propagating metadata is totally ad hoc.

I do want to fix those things, but probably out of scope for this one.

suo committed:
Update on "[export] preserve metadata during nonstrict tracing"


Previously, nonstrict tracing would wipe metadata of graphmodules, because the wrapper class we're using was not detected as a graphmodule and thus meta preservation was not turned on

Differential Revision: [D53139354](https://our.internmc.facebook.com/intern/diff/D53139354/)

[ghstack-poisoned]

thiagocrepaldi commented on torch/export/_trace.py:
```diff
@@ -551,9 +551,14 @@ def __init__(self, mod):
-                        )
+                        if isinstance(self._export_root, torch.fx.GraphModule):
+                            with torch.fx.traceback.preserve_node_meta():
+                                tree_out = torch.fx.Interpreter(self._export_root).run(
```
@titaiwangms @suo the `torch.fx.Interpreter().run()` is used here to preserve `metadata` attribute?

suo replied:
yes, plus: `with torch.fx.traceback.preserve_node_meta():`

suo committed:
Update on "[export] preserve metadata during nonstrict tracing"


Previously, nonstrict tracing would wipe metadata of graphmodules, because the wrapper class we're using was not detected as a graphmodule and thus meta preservation was not turned on

Differential Revision: [D53139354](https://our.internmc.facebook.com/intern/diff/D53139354/)

[ghstack-poisoned]

suo committed:
Update on "[export] preserve metadata during nonstrict tracing"


Previously, nonstrict tracing would wipe metadata of graphmodules, because the wrapper class we're using was not detected as a graphmodule and thus meta preservation was not turned on

Differential Revision: [D53139354](https://our.internmc.facebook.com/intern/diff/D53139354/)

[ghstack-poisoned]

suo committed:
Update on "[export] preserve metadata during nonstrict tracing"


Previously, nonstrict tracing would wipe metadata of graphmodules, because the wrapper class we're using was not detected as a graphmodule and thus meta preservation was not turned on

Differential Revision: [D53139354](https://our.internmc.facebook.com/intern/diff/D53139354/)

[ghstack-poisoned]

suo committed:
Update on "[export] preserve metadata during nonstrict tracing"


Previously, nonstrict tracing would wipe metadata of graphmodules, because the wrapper class we're using was not detected as a graphmodule and thus meta preservation was not turned on

Differential Revision: [D53139354](https://our.internmc.facebook.com/intern/diff/D53139354/)

[ghstack-poisoned]
